src/table_collection.rs

```
@@ -1372,4 +1376,202 @@ impl TableCollection {
+
+        // logicals to indicate whether a site (old) will be kept in new site table
+        let mut keep_sites = vec![false; tables.sites().num_rows().try_into()?];
+
+        let mut last_interval = (Position::from(0.0), Position::from(0.0));
+        for (s, e) in intervals {
+            let (s, e) = (s.into(), e.into());
+            // make sure intervals are sorted
+            if (s > e) || (s < last_interval.1) {
+                return Err(TskitError::RangeError(
+                    "intervals not valid or sorted".into(),
+                ));
+            }
+            keep_sites
+                .iter_mut()
+                .zip(tables.sites_iter())
+                .for_each(|(k, site_row)| {
+                    *k = *k || ((site_row.position >= s) && (site_row.position < e));
+                });
+
+            // use stream_iter and while-let pattern for easier ? operator within a loop
+            let mut edge_iter = tables
+                .edges()
+                .lending_iter()
+                .filter(|edge_row| !((edge_row.right <= s) || (edge_row.left >= e)));
+
+            while let Some(edge_row) = edge_iter.next() {
+                new_edges.add_row_with_metadata(
+                    if edge_row.left < s { s } else { edge_row.left }.into(),
+                    if edge_row.right > e {
+                        e
+                    } else {
+                        edge_row.right
+                    }
+                    .into(),
+                    edge_row.parent.into(),
+                    edge_row.child.into(),
+                    edge_row.metadata.unwrap_or(&[0u8; 0]),
+                )?;
+            }
+
+            let mut migration_iter = tables
+                .migrations()
+                .lending_iter()
+                .filter(|mrow| !!((mrow.right <= s) || (mrow.left >= e)));
+
+            while let Some(migration_row) = migration_iter.next() {
```

The tests do not cover this block.  It may be useful to come up with something to cover this?

---

I left migration table recording out of simulation as it does not work well with simplification as we can see from tskit C api docs for `tsk_table_collection_simplify` (although the `keep_intervals` python code deal with migration table):

> Note
> Migrations are currently not supported by simplify, and an error will be raised if we attempt call simplify on a table collection with greater than zero migrations. See [tskit-dev/tskit#20](https://github.com/tskit-dev/tskit/issues/20)

I can add a note to the Rust `keep_intervals` docs to reflect this.

---

That's fair.  One could envision a test of manually-generated data that  don't require sorting.  But the reality is that that block of code is adding data already present in the tables.  The only way for there to be an error is if someone loaded a table collection with invalid row data generated by another tool.

---

Right. I will make a test that if some one try to call keep_intervals on treeseq that has nonempty migration and set `simplify=true` in the argument, it should return a tskit error.

---

> Before implementing a manual check, just do the test. tskit-c will, I believe, set an error code for that case, so you can just `?` on the simplify call.

Sorry I am not sure I understand your suggestion. Did you mean to do test on  the method that generates a treesequence with non-empty migration table in `src/test_fixtures.rs`  ` generate_simple_treesequence` function ?

I thought I already used `?` for the `simplfy` call in  the `keep_intervals` method for `TableCollection`.  Am I missing something?

---

I think that I misunderstood your comment re: testing.  I think that your tests are actually okay except for the issue of adding the migrations after simplification.